Split the concerns of the commit creation functions

carlosmn opened the pull request:
As discussed on #2681 this PR splits up the commit creation functions into creating the commit and those which update a reference as well.

`_fromstate()` brings us closer to `git-commit` with its defaults for the parents and tree.

I don't love `git_commit_create_on_head()` since you can get the same by passing `"HEAD"` to `git_commit_create_on()` and we already have enough confusion about what we mean by `head` in function names.

The names are whatever I came up with, better names may come up.

ethomson commented on src/commit.c:
```diff
@@ -302,6 +335,79 @@ int git_commit_amend(
+
+	/* TODO: how do we handle octopus merges? */
+	if (state == GIT_REPOSITORY_STATE_MERGE) {
+		if ((error = git_reference_name_to_id(&current_id, repo, GIT_MERGE_HEAD_FILE)) < 0)
```
In the octopus merge case, `MERGE_HEAD` will contain the id of each head being merged one per line.  I suspect that `reference_name_to_id` will fail on this as the file will have what it believes to be trailing garbage.

carlosmn replied:
We should probably use some function that can return a list of ids given MERGE_HEAD instead.

ethomson commented:
I like this a lot, except I agree that the names aren't perfect.

I'm gonna cc a bunch of people in case they have a clever idea.

/cc @nulltoken 
/cc @vmg 
/cc @jamill

ethomson commented:
Then again, re-reading #2681 I am growing increasingly happy with `_create_on`...

carlosmn commented:
To work around the naming issue, and other annoying stuff we have in this part of the code, I think we might want to accept a `git_commit_options` wherein we store the data. Adding small variations otherwise leads to combinatorial explosion or only some variants gaining options.

carlosmn commented:
I've rebased on top of the current master and fixed the issue with having multiple entries in `MERGE_HEAD`. I'm still not sure what we should name all of these things or whether we even want `_fromstate()` to update the reference.

Maybe the ones that do update the ref belong in the `git_repository` namespace? Maybe we only want to have the convenience function that does create the commit act on `HEAD` since it's the only one which has semantics in the git tool?

carlosmn commented:
We now have `_fromstate` which does not touch references and `_fromstate_on_head` which does. I figure this is enough to solve the most common use-cases, and anything more complex you just have to go ahead and do the few steps contained therein yourself.

pks-t commented:
You want to revisit this PR and rebase it before I'm doing a review? Will do so as soon as v0.27.0 is out

carlosmn commented:
Rebased.

pks-t commented on include/git2/commit.h:
```diff
@@ -329,6 +321,26 @@ GIT_EXTERN(int) git_commit_extract_signature(git_buf *signature, git_buf *signed
  *	the given reference will be updated to point to it
  */
 GIT_EXTERN(int) git_commit_create(
```
So, to reiterate the API change here. The only backwards-incompatible change we're doing in this PR is to rename `git_commit_create` to `git_commit_create_on`. As this is a rather central function that is probably in use by a lot of downstream users of libgit2, this is going to break most of them. While the fix is trivial, it is going to be a huge pain e.g. for distro maintainers, who'd have to patch a whole lot of reverse-dependencies of libgit2 to keep them working.

So I'd heavily favour to have a non-breaking API change here. Keep `git_commit_create` as is and instead create a new function that handles creation of a commit without updating a reference. The pain point is that I myself have no idea what to call that function. The following list simply sucks:

- git_commit_create_v2: version numbers are rather generic and don't tell much about the function's changed intention
- git_commit_create_only: the `only` only makes sense if you know the context of updating refs
- git_commit_create_norefupdate: well, this at least says what is different from the other commit function
- git_commit_create_obj: dunno, might be just weird

I think we should either just keep `git_commit_create` untouched or mark it as deprecated. Breaking users without a proper deprecation period and without really good reasons feels irresponsible to me.

ethomson commented:
We've had a few changes since this was opened, to improve our commit APIs. I don't think that it's encompassed all the suggestions here, but it is improved. Closing this as stale.
